### PR TITLE
metrics/probabilities: add quantile score

### LINKED ISSUE #520: probabilistic forecast metrics: add quantile score
This issue is to discuss adding the quantile score to the list of probabilistic forecast metrics.

Given a set of observations `(obs_1, obs_2, ..., obs_n)`, forecasts `(fx_1, fx_2, ..., fx_n})`, and corresponding forecast probabilities `(p_1, p_2, ..., p_n)`, the quantile score is defined as:
- `1/n * sum_{i=1}^n (obs_i - fx_i) * (1{obs_i >= fx_i} - p_i)`

where `1{obs_i >= fx_i}` an indicator function (1 if `obs_i >= fx_i`, 0 otherwise).

The quantile score is similar to the Brier Score (`1/n * sum_{i=1}^n (p_i - 1{obs_i >= fx_i})^2`), but where the Brier Score is unitless (values between `[0, 1]`, the quantile score returns the same units as the observations (e.g. power in MW, then quantile score in MW).

Here's an arbitrary example to illustrate:
- forecast power generation at time t as 4 MW with 50% probability: `fx = 4 MW`, `p = 0.5`
- actual power generation at time t is 5 MW: `obs = 5 MW`
- Brier score = `(0.5 - 1{5 MW >= 4 MW})^2` = `(0.5 - 1)^2` = `0.5^2` = `0.25`
- quantile score: `(5 MW - 4 MW) * (1{5 MW >= 4 MW} - 0.5)` = `(1 MW) * (1 - 0.5)` =` 0.5 MW`

Arguments in favor of adding quantile score:
- the metric is commonly used in quantile regression (dating back to a 1978 paper by Koenker and Bassett)
- the metric is readily computed and compatible with `solarforecastarbiter.metrics.probabilistic` and related code (i.e. there would be minimal code change required to add support)

Arguments against adding the quantile score:
- the metric provides similar information as the Brier score

Attached is the paper from Koenker and Bassett (1978) that introduced quantile regression and a more recent paper from Bouallegue, Pinson and Friederichs (2015) that adds additional discussion of the quantile score.

Koenker and Bassett (1978):
[koenker-bassett.pdf](https://github.com/SolarArbiter/solarforecastarbiter-core/files/4916171/koenker-bassett.pdf)

Bouallegue, Pinson and Friederichs (2015):
[qj.2624.pdf](https://github.com/SolarArbiter/solarforecastarbiter-core/files/4916172/qj.2624.pdf)

### PR DESCRIPTION
**Pull request by dplarson**
Add Quantile Score and Quantile Skill Score for evaluating probabilistic
forecasts.

  - [X] Closes #520 .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

File: solarforecastarbiter/metrics/probabilistic.py [hunk 1]
```diff
@@ -224,7 +318,7 @@ def reliability(obs, fx, fx_prob):
         The reliability of the forecast [unitless], where a perfectly reliable
         forecast has value of 0.
 
-    See Also:
+    See Also
```
Comment (wholmgren) [inline solarforecastarbiter/metrics/probabilistic.py:321]: good catch. remove the extra `-` from the line below too.

**Comment by wholmgren** [comment]
Also we'll need a new rc3 whatsnew file for this (we tagged rc2 today ahead of the test trial initialization).

File: solarforecastarbiter/metrics/probabilistic.py (hunk 1 above)
Comment (dplarson) [reply solarforecastarbiter/metrics/probabilistic.py:321]: Thanks for reminding me. Removing now.

File: solarforecastarbiter/metrics/probabilistic.py [hunk 2]
```diff
@@ -88,6 +88,100 @@ def brier_skill_score(obs, fx, fx_prob, ref, ref_prob):
...
+    -------
+    qs : float
+        The Quantile Score, with the same units as the observations.
+
+    Examples
+    --------
+    >>> obs = 100     # observation [MW]
+    >>> fx = 80       # forecast [MW]
+    >>> fx_prob = 60  # probability [%]
+    >>> quantile_score(obs, fx, fx_prob)   # score [W/m^2]
```
[the remark this answers is not in the document]
Comment (dplarson) [reply solarforecastarbiter/metrics/probabilistic.py]: Thanks for catching this.

File: solarforecastarbiter/metrics/probabilistic.py [hunk 3]
```diff
@@ -88,6 +88,100 @@ def brier_skill_score(obs, fx, fx_prob, ref, ref_prob):
     return skill
 
 
+def quantile_score(obs, fx, fx_prob):
+    """Quantile Score (QS).
+
+        QS = 1/n sum_{i=1}^n (fx_i - obs_i) * (p_i - 1{obs_i > fx_i})
```
[the remark this answers is not in the document]
Comment (dplarson) [reply solarforecastarbiter/metrics/probabilistic.py:94]: Good point. I agree quantile score is meant to be computed for a single probability tau at all times (although the formulation does *technically* enable computing it for multiple probabilities). I'll revise the docstring to reflect this, including adding references (and maybe a `Notes` section).

**Comment by dplarson** [comment]
> Can you also make a PR to the github.io repo with the general metrics documentation?
Sure. I'll finish revising this PR based on your comments and then open a PR to the github.io repo so that we can merge both around the same time.

**Comment by wholmgren** [comment]
@dplarson #535 added a rc3 what's new file so that I could test the readthedocs integration. So you'll need to rebase/merge this PR. I'm hoping that your next push will trigger a doc build for this PR. If not try to close and reopen it.

File: solarforecastarbiter/metrics/probabilistic.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dplarson) [reply solarforecastarbiter/metrics/probabilistic.py]: I've revised the equation, added a Notes section and added references.

**Comment by dplarson** [comment]
The readthedocs integration was working for a while, but now seems to have failed. To make it easier, here's what the doc pages look like for quantile score and quantile skill score (when run locally):

screenshot_doc_qs

screenshot_doc_qss

**Comment by dplarson** [comment]
I *think* I've now covered all the necessary revisions. Also, the PR for adding the QS and QSS to the website is also ready for review: https://github.com/SolarArbiter/SolarArbiter.github.io/pull/162

File: solarforecastarbiter/metrics/tests/test_probabilistic.py [hunk 4]
```diff
@@ -42,6 +42,26 @@ def test_brier_skill_score(fx, fx_prob, ref, ref_prob, obs, value):
...
+def test_quantile_score(obs, fx, fx_prob, value):
+    assert prob.quantile_score(obs, fx, fx_prob) == value
+
+
+@pytest.mark.parametrize("obs,fx,fx_prob,ref,ref_prob,value", [
+    (4, 5, 50, 3, 50, 1 - 0.5 / 0.5),
+    (100, 80, 60, 80, 50, 1 - 8 / 10),
+])
+def test_quantile_skill_score(obs, fx, fx_prob, ref, ref_prob, value):
+    assert prob.quantile_skill_score(obs, fx, fx_prob, ref, ref_prob) == value
```
Comment (wholmgren) [inline solarforecastarbiter/metrics/tests/test_probabilistic.py:62]: Is it worth adding some skill score tests for edge cases?

https://github.com/SolarArbiter/solarforecastarbiter-core/blob/13046c8f73e136d81c35d20436779daeb26f26a8/solarforecastarbiter/metrics/tests/test_deterministic.py#L130-L138

https://github.com/SolarArbiter/solarforecastarbiter-core/pull/395/files#diff-19cf284d1bd4c59fcb5b3fa3dcd65980

Comment (dplarson) [reply solarforecastarbiter/metrics/tests/test_probabilistic.py:62]: I think that's a good idea. Plus it gives an excuse to push a commit to cause the RTD task to re-run.

**Comment by dplarson** [comment]
@wholmgren any other items to review or is this ready to be merged?

**Comment by wholmgren** [comment]
Sorry for overlooking this @dplarson. Can you address this warning? https://github.com/SolarArbiter/solarforecastarbiter-core/pull/530/checks?check_run_id=955308462#step:6:118 Perhaps using the same approach as for the rmse skill score. 

Also makes me think we should consider a more generic skill score function but let's leave that for another day.

**Comment by dplarson** [comment]
> Sorry for overlooking this @dplarson. Can you address this warning?

No problem. I'll address that warning and ping you when ready.

And yes, I do agree a more generic skill score function could be beneficial and is worth discussing later.

**Comment by alorenzo175** [comment]
yes
